sktree/stats/utils.py

```diff
@@ -0,0 +1,246 @@
+    metric: str = "mse",
+    n_repeats: int = 1000,
+    seed: int = None,
+) -> ArrayLike:
+    """Compute null distribution using permutation method.
+
+    Parameters
+    ----------
+    X_test : ArrayLike of shape (n_samples, n_features)
+        The data matrix.
+    y_test : ArrayLike of shape (n_samples, n_outputs)
+        The output matrix.
+    covariate_index : ArrayLike of shape (n_covariates,)
+        The indices of the covariates to permute.
+    est : ForestClassifier
+        The forest that will be used to recompute the test statistic on the permuted data.
+    metric : str, optional
+        The metric, which to compute the null distribution of statistics, by default 'mse'.
+    n_repeats : int, optional
+        The number of times to sample the null, by default 1000.
+    seed : int, optional
+        Random seed, by default None.
+    """
+    rng = np.random.default_rng(seed)
+    X_test, y_test = check_X_y(X_test, y_test, ensure_2d=True, multi_output=True)
+    n_samples_test = len(y_test)
+    n_samples_train = len(y_train)
+    metric_func = METRIC_FUNCTIONS[metric]
+
+    # pre-allocate memory for the index array
+    train_index_arr = np.arange(n_samples_train, dtype=int).reshape(-1, 1)
+    test_index_arr = np.arange(n_samples_test, dtype=int).reshape(-1, 1)
+
+    null_metrics = np.zeros((n_repeats,))
+
+    for idx in range(n_repeats):
+        # permute the covariates inplace
+        rng.shuffle(test_index_arr)
+        perm_X_cov = X_test[test_index_arr, covariate_index]
+        X_test[:, covariate_index] = perm_X_cov
+
+        rng.shuffle(train_index_arr)
+        perm_X_cov = X_train[train_index_arr, covariate_index]
+        X_train[:, covariate_index] = perm_X_cov
+
+        # train a new forest on the permuted data
+        # XXX: should there be a train/test split here? even w/ honest forests?
+        est.fit(X_train, y_train.ravel())
+        y_pred = est.predict(X_test)
+
+        # compute two instances of the metric from the sampled trees
+        metric_val = metric_func(y_test, y_pred)
+
+        null_metrics[idx] = metric_val
+    return null_metrics
+
+
+def _compute_null_distribution_coleman(
+    y_test: ArrayLike,
+    y_pred_proba_normal: ArrayLike,
+    y_pred_proba_perm: ArrayLike,
+    metric: str = "mse",
+    n_repeats: int = 1000,
+    seed: int = None,
+) -> Tuple[ArrayLike, ArrayLike]:
+    """Compute null distribution using Coleman method.
+
+    The null distribution is comprised of two forests.
+
+    Parameters
+    ----------
+    X_test : ArrayLike of shape (n_samples, n_features)
+        The data matrix.
+    y_test : ArrayLike of shape (n_samples, n_outputs)
+        The output matrix.
+    y_pred_proba_normal : ArrayLike of shape (n_samples_normal, n_outputs)
+        The predicted posteriors from the normal forest.
+    y_pred_proba_perm : ArrayLike of shape (n_samples_perm, n_outputs)
+        The predicted posteriors from the permuted forest.
```

Comment:
These should be of shape `(n_trees, n_samples, n_outputs)` @adam2392

Reply:
When sample_dataset_per_tree is on, both arrays could have NaN values.

Reply:
Will review this afternoon, sorry about the delay

Reply:
Resolved. I believe(?)